src/Symfony/Component/DependencyInjection/Tests/Compiler/MergeExtensionConfigurationPassTest.php

```diff
@@ -84,6 +85,175 @@ public function testOverriddenEnvsAreMerged()
+            'baz' => '%env(BAZ)%',
+        );
+
+        $this->assertSame($expected, $container->resolveEnvPlaceholders($ext->getConfig()));
+    }
+
+    public function testConcatenatedEnvInConfig()
+    {
+        $container = new ContainerBuilder();
+        $container->registerExtension($ext = new FooExtension());
+        $container->prependExtensionConfig('foo', array(
+            'bar' => $expected = 'foo %env(BAR)% baz',
+        ));
+
+        (new RegisterEnvVarProcessorsPass())->process($container);
+        (new MergeExtensionConfigurationPass())->process($container);
+
+        $this->assertSame(array('bar' => $expected), $container->resolveEnvPlaceholders($ext->getConfig()));
+    }
+
+    /**
+     * @expectedException \Symfony\Component\Config\Definition\Exception\InvalidConfigurationException
+     * @expectedExceptionMessage A dynamic value is not compatible with a enumeration node at path "foo.enum_node".
+     */
+    public function testEnvIsIncompatibleWithEnumNode()
+    {
+        $container = new ContainerBuilder();
+        $container->registerExtension(new FooExtension());
+        $container->prependExtensionConfig('foo', array(
+            'enum_node' => '%env(FOO)%',
+        ));
+
+        (new RegisterEnvVarProcessorsPass())->process($container);
+        (new MergeExtensionConfigurationPass())->process($container);
+    }
+
+    /**
+     * @expectedException \Symfony\Component\Config\Definition\Exception\InvalidConfigurationException
+     * @expectedExceptionMessage A dynamic value is not compatible with an array node at path "foo.simple_array_node".
+     */
+    public function testEnvIsIncompatibleWithArrayNode()
+    {
+        $container = new ContainerBuilder();
+        $container->registerExtension(new FooExtension());
+        $container->prependExtensionConfig('foo', array(
+            'simple_array_node' => '%env(json:FOO)%',
+        ));
+
+        (new RegisterEnvVarProcessorsPass())->process($container);
+        (new MergeExtensionConfigurationPass())->process($container);
+    }
+
+    public function testNormalizedEnvIsCompatibleWithArrayNode()
+    {
+        $container = new ContainerBuilder();
+        $container->registerExtension($ext = new FooExtension());
+        $container->prependExtensionConfig('foo', array(
+            'array_node' => $expected = '%env(CHILD)%',
+        ));
+
+        (new RegisterEnvVarProcessorsPass())->process($container);
+        (new MergeExtensionConfigurationPass())->process($container);
+
+        $this->assertSame(array('array_node' => array('child_node' => $expected)), $container->resolveEnvPlaceholders($ext->getConfig()));
+    }
+
+    public function testEnvCanBeUnset()
+    {
+        $container = new ContainerBuilder();
+        $container->registerExtension($ext = new FooExtension());
+        $container->prependExtensionConfig('foo', array(
+            'array_node' => $expected = array('int_unset_at_zero' => '%env(int:CHILD)%'),
```

Comment:
@nicolas-grekas think we need to revert this. Value might not be zero during runtime.